File: design-pattern-python/design_pattern_python/bridge/display.py
from display_impl import DisplayImpl, StringDisplayImpl
from typing_extensions import final


class Display:
    def __init__(self, impl: DisplayImpl) -> None:
        self.__impl = impl

    def open(self) -> None:
        self.__impl.raw_open()

    def print(self) -> None:
        self.__impl.raw_print()

    def close(self) -> None:
        self.__impl.raw_close()

    @final
    def display(self) -> None:
        self.open()
        self.print()
        self.close()


class CountDisplay(Display):
    def multi_display(self, times: int) -> None:
        self.open()
        for _ in range(times):
            self.print()
        self.close()


if __name__ == "__main__":
    d1 = Display(StringDisplayImpl("Hello, Japan."))
    d2 = CountDisplay(StringDisplayImpl("Hello, World."))
    d3 = CountDisplay(StringDisplayImpl("Hello, Universe."))
    d1.display()
    d2.display()
    d3.display()
    d3.multi_display(5)

File: design-pattern-go/bridge/display/display.go
package display

import (
	di "design-pattern-go/bridge/display_impl"
)

type Display struct {
	impl di.DisplyImpl
}

func NewDisplay(impl di.DisplyImpl) *Display {
	return &Display{
		impl,
	}
}

func (d *Display) Open() {
	d.impl.RawOpen()
}

func (d *Display) Print() {
	d.impl.RawPrint()
}

func (d *Display) Close() {
	d.impl.RawClose()
}

func (d *Display) Display() {
	d.Open()
	d.Print()
	d.Close()
}

type CountDisplay struct {
	Display
}

func NewCountDisplay(impl di.DisplyImpl) *CountDisplay {
	return &CountDisplay{
		*NewDisplay(impl),
	}
}

func (c *CountDisplay) MultiDisplay(times int) {
	c.Open()
	for i := 0; i < times; i++ {
		c.Print()
	}
	c.Close()
}
